### PR TITLE
feat(Menu): Make Submenu open state controlled

### PR DESCRIPTION
**Pull request by sophieH29**
This PR aims make submenu open  state controlled, thus to bring the flexibility of using submenu and rewrite default behavior.

### CODE REVIEW

File: packages/react/src/lib/AutoControlledComponent.tsx [hunk 1]
```diff
@@ -187,8 +187,9 @@ export default class AutoControlledComponent<P = {}, S = {}> extends UIComponent
    * Second argument is a state object that is always passed to setState.
    * @param {object} maybeState State that corresponds to controlled props.
    * @param {object} [state] Actual state, useful when you also need to setState.
+   * @param {object} onStateChanged Callback which is called after setState applied.
    */
-  trySetState = (maybeState, state?) => {
+  trySetState = (maybeState, state?, onStateChanged?: () => void) => {
```
Comment (layershifter) [inline packages/react/src/lib/AutoControlledComponent.tsx]: https://github.com/Semantic-Org/Semantic-UI-React/pull/3452#discussion_r264972028

I support there Levi, I think that we can also remove `state` param, too. And match React's signature:

```suggestion
  trySetState = (state: Partial<State>, callback?: () => void) => {
```

File: packages/react/src/components/Menu/MenuItem.tsx [hunk 2]
```diff
@@ -363,12 +372,20 @@ class MenuItem extends AutoControlledComponent<ReactProps<MenuItemProps>, MenuIt
...
       e.stopPropagation()
       e.preventDefault()
     }
   }
+
+  private trySetMenuOpen(newValue: boolean, eventArgs: any, onStateChanged?: any) {
+    this.trySetState({ menuOpen: newValue }, null, onStateChanged)
+    _.invoke(this.props, 'onMenuOpenChange', eventArgs, {
+      ...this.props,
+      ...{ menuOpen: newValue },
```
Comment (layershifter) [inline packages/react/src/components/Menu/MenuItem.tsx]:
```suggestion
      menuOpen: newValue,
```

File: packages/react/src/lib/AutoControlledComponent.tsx (hunk 1 above)
Comment (sophieH29) [reply packages/react/src/lib/AutoControlledComponent.tsx]: I like that, actually didn't find any usage of 2nd param.
Would make this change if nobody minds

File: packages/react/src/lib/AutoControlledComponent.tsx [hunk 3]
```diff
@@ -187,13 +187,14 @@ export default class AutoControlledComponent<P = {}, S = {}> extends UIComponent
    * Second argument is a state object that is always passed to setState.
    * @param {object} maybeState State that corresponds to controlled props.
    * @param {object} [state] Actual state, useful when you also need to setState.
+   * @param {object} callback Callback which is called after setState applied.
    */
-  trySetState = (maybeState, state?) => {
+  trySetState = (state: Partial<S>, callback?: () => void) => {
     const { autoControlledProps } = this.constructor as any
     if (process.env.NODE_ENV !== 'production') {
```
Comment (layershifter) [inline packages/react/src/lib/AutoControlledComponent.tsx]: This check should be removed, otherwise we will get incorrect warnings

File: packages/react/src/components/Menu/MenuItem.tsx [hunk 4]
```diff
@@ -362,12 +372,25 @@ class MenuItem extends AutoControlledComponent<ReactProps<MenuItemProps>, MenuIt
...
+    // The reason why post-effect is not passed as callback to trySetState method
+    // is that in 'controlled' mode the post-effect is applied before final re-rendering
+    // which cause a broken behavior: for e.g. when it is needed to focus submenu trigger on ESC.
+    // TODO: all DOM post-effects should be applied at componentDidMount & componentDidUpdated stages.
+    onStateChanged && onStateChanged()
+    _.invoke(this.props, 'onMenuOpenChange', eventArgs, {
+      ...this.props,
+      menuOpen: newValue,
+    })
+  }
```
Comment (layershifter) [inline packages/react/src/components/Menu/MenuItem.tsx:392]: I don't like this method, but I don't have a better option now...

File: docs/src/examples/components/Menu/Usages/MenuExampleWithSubmenuControlled.shorthand.tsx [hunk 5]
```diff
@@ -0,0 +1,50 @@
...
+            {
+              key: '2',
+              content: 'item2',
+              menu: [{ key: '1', content: 'item2.1' }, { key: '2', content: 'item2.2' }],
+            },
+            {
+              key: '3',
+              content: 'item3',
+              menu: [{ key: '1', content: 'item3.1' }, { key: '2', content: 'item3.2' }],
+            },
```
Comment (layershifter) [inline docs/src/examples/components/Menu/Usages/MenuExampleWithSubmenuControlled.shorthand.tsx]: I suggest to make it more compact:

```
menu: [
  { key: '1', content: 'item1' },
  { key: '2', content: 'item2' },
  { key: '3', content: 'item3' },
],
```

File: docs/src/examples/components/Menu/Usages/MenuExampleWithSubmenuControlled.shorthand.tsx [hunk 6]
```diff
@@ -0,0 +1,50 @@
...
+    ]
+  }
+
+  render() {
+    return (
+      <>
+        <span>{`Editorials menu item requested to change its submenu open state to "${
+          this.state.editorialsMenuOpen
+        }".`}</span>
+        <Menu defaultActiveIndex={0} items={this.renderItems()} />
```
Comment (layershifter) [inline docs/src/examples/components/Menu/Usages/MenuExampleWithSubmenuControlled.shorthand.tsx]: This looks unreadable 🤔 

What about something like this:

```jsx
<Header as="h5" content="Current state:" />
<pre>{JSON.stringify(this.state, null, 2)}</pre>
<Divider />
<Menu defaultActiveIndex={0} items={this.renderItems()} />
  
```

File: docs/src/examples/components/Menu/Usages/MenuExampleWithSubmenuControlled.shorthand.tsx (hunk 5 above)
Comment (layershifter) [reply docs/src/examples/components/Menu/Usages/MenuExampleWithSubmenuControlled.shorthand.tsx]: As it will become more compact, do we need `renderItems()` at all?

Comment (sophieH29) [reply docs/src/examples/components/Menu/Usages/MenuExampleWithSubmenuControlled.shorthand.tsx]: I would leave `renderItems()`, for better readability

File: docs/src/examples/components/Popup/Types/PopupExample.shorthand.tsx [hunk 7]
```diff
@@ -1,6 +1,19 @@
...
+    content={
+      <Dialog
+        cancelButton="Cancel"
+        confirmButton="Confirm"
+        header="Action confirmation"
+        trigger={<Button content="Open a dialog" />}
+      />
+    }
+  />
+)
```
Comment (layershifter) [inline docs/src/examples/components/Popup/Types/PopupExample.shorthand.tsx]: Should we revert these changes?

File: packages/react/src/components/Menu/MenuItem.tsx [hunk 8]
```diff
@@ -362,12 +372,25 @@ class MenuItem extends AutoControlledComponent<ReactProps<MenuItemProps>, MenuIt
...
     if (menu && !menuOpen) {
-      this.setState({ menuOpen: true })
+      this.trySetMenuOpen(true, e)
       _.invoke(this.props, 'onActiveChanged', e, { ...this.props, active: true })
       e.stopPropagation()
       e.preventDefault()
     }
   }
+
+  private trySetMenuOpen(newValue: boolean, eventArgs: any, onStateChanged?: any) {
```
Comment (layershifter) [inline packages/react/src/components/Menu/MenuItem.tsx]:
```suggestion
  private trySetMenuOpen(e: React.SyntethicEvent, newValue: boolean, onStateChanged?: any) {
```

nit: Can we place `e` param on the first place as in other methods?

File: packages/react/src/lib/AutoControlledComponent.tsx [hunk 9]
```diff
@@ -187,37 +187,17 @@ export default class AutoControlledComponent<P = {}, S = {}> extends UIComponent
...
-      if (autoControlledProps.indexOf(prop) === -1) return acc
-
       acc[prop] = maybeState[prop]
       return acc
     }, {})
 
-    if (state) newState = { ...newState, ...state }
-
-    if (Object.keys(newState).length > 0) this.setState(newState)
+    if (Object.keys(newState).length > 0) this.setState(newState, callback)
```
Comment (layershifter) [inline packages/react/src/lib/AutoControlledComponent.tsx:201]: Not sure, that we really need this condition, but let's leave it as is for now

**Review by layershifter — APPROVED** [review]

File: packages/react/src/components/Menu/MenuItem.tsx (hunk 8 above)
Comment (sophieH29) [reply packages/react/src/components/Menu/MenuItem.tsx]: fyi, the order is the same as in Popup for `trySetOpen` method. 
And the other thing, if you would want to change state without event, you would have to call
`trySetMenuOpen(null, newValue)`
instead of `trySetMenuOpen(newValue)`. I would leave an order as it is as I think the `newValue` param has biggest priority than `event` object

File: docs/src/examples/components/Popup/Types/PopupExample.shorthand.tsx (hunk 7 above)
Comment (sophieH29) [reply docs/src/examples/components/Popup/Types/PopupExample.shorthand.tsx]: oops, yeah. Did some testing, great you noticed
